src/lib.rs

```diff
@@ -3,47 +3,76 @@
 pub extern crate rustls;
 pub extern crate webpki;
 
-extern crate futures;
-extern crate tokio_io;
 extern crate bytes;
+extern crate futures;
 extern crate iovec;
+extern crate tokio_io;
 
-mod common;
 pub mod client;
+mod common;
 pub mod server;
 
-use std::{ io, mem };
+use common::Stream;
+use futures::{Async, Future, Poll};
+use rustls::{ClientConfig, ClientSession, ServerConfig, ServerSession};
 use std::sync::Arc;
+use std::{io, mem};
+use tokio_io::{try_nb, AsyncRead, AsyncWrite};
 use webpki::DNSNameRef;
-use rustls::{
-    ClientSession, ServerSession,
-    ClientConfig, ServerConfig
-};
-use futures::{Async, Future, Poll};
-use tokio_io::{ AsyncRead, AsyncWrite, try_nb };
-use common::Stream;
 
+#[derive(Debug, Copy, Clone)]
+pub enum TlsState {
+    #[cfg(feature = "early-data")]
+    EarlyData,
+    Stream,
+    ReadShutdown,
+    WriteShutdown,
+    FullyShutdown,
+}
+
+impl TlsState {
+    pub(crate) fn shutdown_read(&mut self) {
+        match *self {
+            TlsState::WriteShutdown | TlsState::FullyShutdown => *self = TlsState::FullyShutdown,
+            _ => *self = TlsState::ReadShutdown,
```

Comment:
This may cause early data to not be written?

Reply:
not sure I follow. Do you mean early data won't write out if we shutdown the read half when it's in EARLY_DATA stage?

Reply:
Yes.

Reply:
> Do you mean early data won't write out if we shutdown the read half when it's in EARLY_DATA stage?

There should be a test that would have caught this. Is it practical to create such a test?

Reply:
> There should be a test that would have caught this. Is it practical to create such a test?

This is a bit difficult because we need to build a server that supports early data, and rustls currently only supports clients.

Reply:
hmmm, does it make sense to have a TCP server which just tests if early data has been properly sent out?

Reply:
> hmmm, does it make sense to have a TCP server which just tests if early data has been properly sent out?

The scenario here is that the server refuses to receive data via early data, at which point `tokio-rustls` will fallback to normal delivery.

We need a working TLS 1.3 server, and reject 0-RTT data when client makes a second connection.

Reply:
I seem to be thinking too much. It is not executed to `shutdown_read` during the early data phase. :(